### ISSUE TITLE
[Windows] Missing scrollbars in compare editor on non default screen scaling

### ISSUE THREAD

**Issue by merks** [eclipse-platform/eclipse.platform#1489]
Let's make sure issue is not already fixed in latest builds first.

- [x] I verified I can reproduce this issue against [latest Integration Build of Eclipse SDK](https://download.eclipse.org/eclipse/downloads/)

## Steps to reproduce

From a fresh installation and clean workspace:

Any files I compare in my updated SDK IDE, it doesn't show scroll bars:

![image](https://github.com/user-attachments/assets/a6fb9ee8-ef88-4fcb-9d59-57ea3c9121a3)

But in a debug launched IDE it works.

I noticed this in other recently installed/updated IDEs.

## Tested under this environment:

I'm on Windows 10.

**Comment by iloveeclipse** [comment]
@merks : could you please disable themes in preferences and check if that would "fix" the diff editor?
@BeckerWdf : could it be related to latest theme changes?

**Comment by merks** [comment]
Neither no theme, nor the choice of theme has an impact.   I tried clicking around in case it was just not visibly painted but actually there, but it really seems not to be there at all, just the blank area where it should be.

**Comment by iloveeclipse** [comment]
I see it now too. Bisecting...
SDK-I20240612-0510 OK
SDK-I20240725-0200 BAD

**Comment by iloveeclipse** [comment]
I20240618-1800 BAD => Commit https://github.com/eclipse-platform/eclipse.platform.swt/commit/25b32ef8091c278928027b3326a83033874f134b from https://github.com/eclipse-platform/eclipse.platform.swt/pull/1209 is the bad one.

@ShahzaibIbrahim, @fedejeanne, the PR above broke compare editor on Windows.
This is a blocker for 4.33, as it makes the compare editor hardly usable. 

Please investigate as soon as possible!

**Comment by ShahzaibIbrahim** [comment]
Can I have more info on this? I tried reproducing on latest build with no VM arguments, Windows 11, Screen Resolution 1920x1080 and zoom level 100%. 

I am unable to reproduce it for now. Perhaps, I am missing something?

![image](https://github.com/user-attachments/assets/2b7b4ed9-e5be-448e-88bf-97b32a99d63d)

**Comment by iloveeclipse** [comment]
Where one can find zoom value on Windows? I have Notebook with Win 11 and 1920x1080 resolution too.

**Comment by ShahzaibIbrahim** [comment]
Systems -> Display -> Scale

![image](https://github.com/user-attachments/assets/5ebf594f-4326-4c18-a0c4-76cdb6571d01)

**Comment by iloveeclipse** [comment]
OK, I have 125%

**Comment by iloveeclipse** [comment]
After switching to 100% I see the tiny line for the scrollbar now, and it shows on hover, but I can't read the text anymore because it is all so small now :-)

**Comment by merks** [comment]
My primary display is also 125%  The secondary is 225%.

**Comment by ShahzaibIbrahim** [comment]
It's now reproduced. It has indeed raised the regression. I will need time to investigate but in the meantime, I can either
1) Revert the change completely
2) or fix the regression with a small change but it won't fix the original issue entirely.

Kindly suggest.

**Comment by iloveeclipse** [comment]
4.33 M3 build is planned for 15 August.

The fix (or revert) should be done before, so we could deliver M3 for testing and could revert the fix if it would be still not OK.

**Comment by ShahzaibIbrahim** [comment]
I will create a new PR for this fix, and will address my issue later.

**Comment by merks** [comment]
Thanks folks!

**Comment by HeikoKlare** [comment]
To me, it look as if the introduced method simply uses the wrong zoom value.
[This line](https://github.com/eclipse-platform/eclipse.platform.swt/blob/1984e06f3cd75c7f5d1329721d3bcfc2b8152bb4/bundles/org.eclipse.swt/Eclipse%20SWT/win32/org/eclipse/swt/widgets/Widget.java#L2672):
```java
return OS.GetSystemMetricsForDpi(nIndex, DPIUtil.mapZoomToDPI(getZoom()));
```
should instead be:
```java
return OS.GetSystemMetricsForDpi(nIndex, DPIUtil.mapZoomToDPI(nativeZoom));
```
That's why the issue only appear at 125% (and other quarter scales): `getZoom()` return 100% instead of the proper 125%.

In preliminary tests, this fixes the problem for me.

@ShahzaibIbrahim can you please further investigate whether that might be the cause?

**Comment by ShahzaibIbrahim** [comment]
> To me, it look as if the introduced method simply uses the wrong zoom value. [This line](https://github.com/eclipse-platform/eclipse.platform.swt/blob/1984e06f3cd75c7f5d1329721d3bcfc2b8152bb4/bundles/org.eclipse.swt/Eclipse%20SWT/win32/org/eclipse/swt/widgets/Widget.java#L2672):
> 
> ```java
> return OS.GetSystemMetricsForDpi(nIndex, DPIUtil.mapZoomToDPI(getZoom()));
> ```
> 
> should instead be:
> 
> ```java
> return OS.GetSystemMetricsForDpi(nIndex, DPIUtil.mapZoomToDPI(nativeZoom));
> ```
> 
> That's why the issue only appear at 125% (and other quarter scales): `getZoom()` return 100% instead of the proper 125%.
> 
> In preliminary tests, this fixes the problem for me.
> 
> @ShahzaibIbrahim can you please further investigate whether that might be the cause?

You are right, this is exactly the issue. The higher the zoom level goes, the smaller scrollbar became because getZoom() returned 100% in case the zoom is 125% or 150% and 200% in case of 175%. Using nativeZoom has fixed the issue (I checked in all zoom levels)

**Comment by iloveeclipse** [comment]
Is this issue fixed? If yes, please close.

**Comment by HeikoKlare** [comment]
Has been fixed via:
- eclipse-platform/eclipse.platform.swt#1397

I revalidated with latest I-Build (I20240826-2120): the original issue is not present
